jfr-daemon/src/main/java/com/newrelic/jfr/daemon/JFRJMXRecorder.java

```diff
@@ -34,14 +36,55 @@ public JFRJMXRecorder(
     this.streamFromJmx = streamFromJmx;
   }
 
+  public static JFRJMXRecorder connectWithBackOff(DaemonConfig config) throws IOException {
+    JFRJMXRecorder out = null;
+    try {
+      out = connect(config);
+    } catch (IOException e) {
+      for (int i = 0; i < backoffSeconds.size(); i = i + 1) {
+        try {
+          out = connect(config);
+          break;
+        } catch (IOException iox) {
+          // Log retry?
+          if (i == backoffSeconds.size() - 1) {
+            throw iox;
+          }
+        }
+      }
+    }
+
+    return out;
+  }
+
+  public void startRecordingWithBackOff()
+      throws MalformedObjectNameException, ReflectionException, IOException,
+          InstanceNotFoundException, MBeanException, OpenDataException {
+    try {
+      startRecording();
+    } catch (Exception e) {
+      for (int i = 0; i < backoffSeconds.size(); i = i + 1) {
```

Comment:
I would prefer not to see the same construct repeated.  The recursive approach might help, but maybe it warrants an abstraction?

Reply:
I, too, would prefer that this construct not be repeated. However, I had a look at it and I suspect that abstracting it is trickier than it appears (e.g. due to differences in exception signature) and the resulting code is going to look damned ugly and hard to read. This is the kind of situation where I wish we had more of a functional language to work in - if this was Scala (or Kotlin?) then this would become a "with" pattern, but we are where we are.

This was quite a long-winded way of saying that I agree, but in the interests of getting this feature out to users, I've gone with the repeated construct for now. I'm up for a tidied-up functionalist version if it's not too complex or ugly, but let's get this in sooner.